src/dune_rules/melange_rules.ml

```diff
@@ -16,36 +16,48 @@ let ocaml_flags sctx ~dir melange =
     let ocaml_version = (Super_context.context sctx).version in
     Super_context.with_vendored_flags ~ocaml_version flags
 
-let lib_output_dir ~target_dir ~lib_dir =
+let lib_output_dir ~sctx ~target_dir lib =
+  let info = Lib.info lib in
+  let lib_dir =
+    match Lib_info.status info with
+    | Private _ ->
+      let lib = Lib.Local.of_lib_exn lib in
+      let info = Lib.Local.info lib in
+      Lib_info.src_dir info
+    | Public _ ->
+      let package_name = Option.value_exn (Lib_info.package info) in
+      let bctx = (Super_context.context sctx).build_dir in
+      let info = Lib.info lib in
+      let src_dir = Lib_info.src_dir info in
+      Path.Build.L.relative bctx
+        [ "node_modules"
+        ; Package.Name.to_string package_name
+        ; Path.Source.to_string (Path.drop_build_context_exn src_dir)
+        ]
+    | Installed | Installed_private ->
+      let package_name = Option.value_exn (Lib_info.package info) in
+      let bctx = (Super_context.context sctx).build_dir in
+      Path.Build.L.relative bctx
+        [ "node_modules"; Package.Name.to_string package_name ]
```

Comment:
`Public` is handled differently than `| Installed | Installed_private` which means that installation is going to be broken. I think for `Public` libraries you need to remove the `src_dir` component. If a library is public, the user can't make assumptions as to where its located